feat(dpg): optional properties

archerzz opened the pull request:
- add `IsRequired` property to `FieldDeclaration`
- update constructor parameter list, only add to public constructor parameters if it's required and not readonly
- update conditions to determine readonly properties
- update model writer, use change tracking types for optional collection properties
- add new test cases
- fix old test cases which contains optional property definitions
- update test project

resolve #2339
resolve #2490

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

AlexanderSher commented on src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:
```diff
@@ -81,33 +82,33 @@ private static (IReadOnlyDictionary<FieldDeclaration, InputModelProperty> Fields
 
-                var field = new FieldDeclaration($"{inputModelProperty.Description}", fieldModifiers, fieldType, inputModelProperty.Name.FirstCharToUpperCase(), writeAsProperty: true);
+                var field = new FieldDeclaration($"{inputModelProperty.Description}", fieldModifiers, fieldType, inputModelProperty.Name.FirstCharToUpperCase(),
+                    isRequired: inputModelProperty.IsRequired, writeAsProperty: true);
```
Is there a difference between read-only and required field?

archerzz replied:
> Is there a difference between read-only and required field?

@AlexanderSher There is. See the [example](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.cs) regarding read-only required/optional fields.
A read-only value type field is different when it's required or not. So we need a flag to indicate whether the field is required or not.
- required: defined as `T`
- optional: defined as `T?`

AlexanderSher commented on src/AutoRest.CSharp/LowLevel/Generation/LowLevelModelWriter.cs:
```diff
@@ -48,9 +49,19 @@ private static void WritePublicConstructor(CodeWriter writer, ConstructorSignatu
+                foreach (var field in model.Fields.Where(f => !initializedFields.Contains(f)))
                 {
-                    writer.Line($"{field.Name:I} = new List<{field.Type.Arguments[0]}>(0).AsReadOnly();");
+                    if (TypeFactory.IsList(field.Type))
```
I think we should use `FieldDeclaration.DefaultValue` here, and `DefaultValue` itself should be set in `ModelTypeProvider`

AlexanderSher replied:
Then the whole block would be something like:
```cs
foreach (var field in model.Fields.Where(f => !initializedFields.Contains(f)))
{
    if (field.DefaultValue is not null) 
    {
        writer.Line($"{field.Name:I} = {field.DefaultValue}");
    }
}
```

archerzz replied:
I tried `DefaultValue`, and the problem is that it will generate something like below:
```csharp
public IList<string> OptionalStringList { get; } = new ChangeTrackingList<string>();
```
The relevant codes are https://github.com/archerzz/autorest.csharp/blob/34cbccd14ea91fc645dfa1b23f9e8b9e8d924c89/src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs#L113

AlexanderSher replied:
Update that line in CodeWriterExtensions.cs to `if ((modifiers.HasFlag(FieldModifiers.Const) || modifiers.HasFlag(FieldModifiers.Static)) && field.DefaultValue != null)`. We should not initialize instance fields this way.

AlexanderSher commented on src/AutoRest.CSharp/LowLevel/Generation/LowLevelModelWriter.cs:
```diff
@@ -48,9 +49,19 @@ private static void WritePublicConstructor(CodeWriter writer, ConstructorSignatu
+                    {
+                        if (TypeFactory.IsReadOnlyList(field.Type))
+                        {
+                            writer.Line($"{field.Name:I} = new List<{field.Type.Arguments[0]}>(0).AsReadOnly();");
```
Why can't we use `Array.Empty<field.Type.Arguments[0]>()` here?

archerzz replied:
I guess it's due to the performance: https://www.tabsoverspaces.com/233833-best-way-to-create-an-empty-collection-array-and-list-in-csharp-net

AlexanderSher replied:
According to this article, `Array.Empty` is the best (which is obvious cause it is a static field). To be clear, the output line should look like this: `OptionalStringList = Array.Empty<string>();`

AlexanderSher commented on src/AutoRest.CSharp/LowLevel/Generation/LowLevelModelWriter.cs:
```diff
@@ -48,9 +49,19 @@ private static void WritePublicConstructor(CodeWriter writer, ConstructorSignatu
+                        }
+                        else if (!field.IsRequired)
+                        {
+                            writer.Line($"{field.Name:I} = {Constant.NewInstanceOf(TypeFactory.GetPropertyImplementationType(field.Type)).GetConstantFormattable()};");
```
I think list and dictionary properties should always have default values. That way we simplify logic for constructor writers.

archerzz commented:
change since last update:
- revert change in `FieldDeclarartion` to remove `IsRequired` property
- calculate the `DefaultValue` for read-only and optional list types
- change model writer logic to generate initialization codes for properties with `DefaultValue`
- update `CodeWriterExtension` to skip default value when generating field declaration codes for non-static or non-const fields
- replace `new List<>(0).AsReadOnly()` with `Array.Empty<>()` to initialize read-only lists

For the last item, according to the benchmark test, using `Array.Empty<string>()` is the fastest.
|         Method |       Mean |     Error |    StdDev |    Median |  Gen 0 | Allocated |
|--------------- |-----------:|----------:|----------:|----------:|-------:|----------:|
| FromArrayEmpty |  0.0367 ns | 0.0373 ns | 0.0719 ns | 0.0000 ns |      - |         - |
|   FromListZero | 10.0773 ns | 0.2939 ns | 0.6135 ns | 9.9391 ns | 0.0130 |      56 B |
|       FromList |  9.7521 ns | 0.2816 ns | 0.3458 ns | 9.7247 ns | 0.0130 |      56 B |

Benchmark codes:
```csharp
public class EmptyListBenchmark
    {
        [Benchmark]
        public IReadOnlyList<string> FromArrayEmpty()
        {
            return Array.Empty<string>();
        }

        [Benchmark]
        public IReadOnlyList<string> FromListZero()
        {
            return new List<string>(0).AsReadOnly();
        }

        [Benchmark]
        public IReadOnlyList<string> FromList()
        {
            return new List<string>().AsReadOnly();
        }
    }
```

archerzz commented on src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:
```diff
@@ -81,33 +82,33 @@ private static (IReadOnlyDictionary<FieldDeclaration, InputModelProperty> Fields
                 var propertyIsRequiredInNonRoundTripModel = (inputModel.Usage is InputModelTypeUsage.Input or InputModelTypeUsage.Output) && inputModelProperty.IsRequired;
-                var propertyIsReadOnly = inputModelProperty.IsReadOnly || propertyIsCollection || propertyIsRequiredInNonRoundTripModel;
+                var propertyIsOptionalInOutputModel = (inputModel.Usage is InputModelTypeUsage.Output) && !inputModelProperty.IsRequired;
+                var propertyIsReadOnly = inputModelProperty.IsReadOnly || propertyIsCollection || propertyIsRequiredInNonRoundTripModel || propertyIsOptionalInOutputModel;
```
> I don't quite understand for the definition of propertyIsRequiredInNonRoundTripModel, if a property is required in an InputModel, how can we say this is readonly?

The required property must be set in the constructor. After that, you cannot change it. See https://github.com/Azure/autorest.csharp/issues/2451
`

AlexanderSher commented on src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:
```diff
@@ -120,9 +121,30 @@ private static CSharpType GetDefaultPropertyType(in InputModelTypeUsage modelUsa
+                }
+                if (!isRequired)
+                {
+                    return $"{Constant.NewInstanceOf(TypeFactory.GetPropertyImplementationType(propertyType)).GetConstantFormattable()};";
```
`GetConstantFormattable()` returns `FormattableString`, so there is no need to wrap it into another formattable:
```cs
return Constant.NewInstanceOf(TypeFactory.GetPropertyImplementationType(propertyType)).GetConstantFormattable();
```

AlexanderSher commented on src/AutoRest.CSharp/LowLevel/Generation/LowLevelModelWriter.cs:
```diff
@@ -48,9 +49,12 @@ private static void WritePublicConstructor(CodeWriter writer, ConstructorSignatu
                 var initializedFields = WriteFieldsInitialization(writer, signature, model);
                 // TODO: Add IReadOnlyDictionary
-                foreach (var field in model.Fields.Where(f => !initializedFields.Contains(f) && TypeFactory.IsReadOnlyList(f.Type)))
+                foreach (var field in model.Fields.Where(f => !initializedFields.Contains(f)))
```
TODO about `IReadOnlyDictionary` should be moved into ModelTypeProvider as well

AlexanderSher reviewed (approved):
Couple of small non-blocking comments. Needs approval from @annelo-msft for updated output.

annelo-msft reviewed (commented):
Would it be possible to add generated output models that have optional properties and properties that are collections of models?  It would be helpful to have examples showing the output of the generator for the features that are being added.  Thanks!

archerzz commented:
> Would it be possible to add generated output models that have optional properties and properties that are collections of models? It would be helpful to have examples showing the output of the generator for the features that are being added. Thanks!

added.

change since last update:
- remove unnecessary formattable string declaration
- remove tailing `;` in the `DefaultValue`
- move TODO comment
- add and refactor test case

AlexanderSher commented:
> properties that are collections of models?

Here are collections of models: 
- https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Models-Cadl/Generated/CollectionItem.cs
- https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Models-Cadl/Generated/CollectionItem.Serialization.cs#L22
